runtime/connections.go

```diff
@@ -5,48 +5,65 @@ import (
 	"fmt"
 	"strings"
 
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
 	"github.com/rilldata/rill/runtime/drivers"
 	"github.com/rilldata/rill/runtime/services/catalog"
 )
 
-func (r *Runtime) newMetaStore(ctx context.Context, instanceID string) (drivers.Handle, func(), error) {
-	c, err := r.ConnectorDefByName(r.opts.MetastoreDriver)
-	if err != nil {
-		panic(err)
+func (r *Runtime) AcquireGlobalHandle(ctx context.Context, connector string) (drivers.Handle, func(), error) {
+	for _, c := range r.opts.SystemConnectors {
+		if c.Name == connector {
+			return r.connCache.get(ctx, "", c.Type, r.connectorConfig(r.opts.MetastoreDriver, c.Config, nil), true)
+		}
 	}
-
-	return r.connCache.get(ctx, instanceID, c.Type, r.variables(r.opts.MetastoreDriver, c.Configs, nil), true)
+	return nil, nil, fmt.Errorf("connector %s doesn't exist", connector)
 }
 
-func (r *Runtime) Registry() drivers.RegistryStore {
-	registry, ok := r.metastore.AsRegistry()
-	if !ok {
-		// Verified as registry in New, so this should never happen
-		panic("metastore is not a registry")
+// AcquireHandle returns instance specific handle
+func (r *Runtime) AcquireHandle(ctx context.Context, instanceID, connector string) (drivers.Handle, func(), error) {
+	instance, err := r.FindInstance(ctx, instanceID)
+	if err != nil {
+		return nil, nil, err
 	}
-	return registry
+
+	if c, err := r.connectorDef(instance, connector); err == nil {
+		return r.connCache.get(ctx, instanceID, c.Type, r.connectorConfig(connector, c.Config, instance.ResolveVariables()), false)
+	}
+
+	// neither defined in rill.yaml nor set in instance, directly used in source
+	return r.connCache.get(ctx, instanceID, connector, r.connectorConfig(connector, nil, instance.ResolveVariables()), false)
 }
 
-func (r *Runtime) AcquireHandle(ctx context.Context, instanceID, connector string) (drivers.Handle, func(), error) {
+// FlushHandle flushes the db handle for the specific connector from the cache
+func (r *Runtime) FlushHandle(ctx context.Context, instanceID, connector string, drop bool) error {
 	instance, err := r.FindInstance(ctx, instanceID)
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
 
-	if instance.RillYAML != nil {
-		// defined in rill.yaml
-		for _, c := range instance.RillYAML.Connectors {
-			if c.Name == connector {
-				return r.connCache.get(ctx, instanceID, c.Type, r.variables(connector, c.Configs, instance.ResolveVariables()), false)
-			}
-		}
+	var driverType string
+	var connectorConfig map[string]any
+	if c, err := r.connectorDef(instance, connector); err == nil {
+		driverType = c.Type
+		connectorConfig = r.connectorConfig(connector, c.Config, instance.ResolveVariables())
+	} else {
+		driverType = connector
+		connectorConfig = r.connectorConfig(connector, nil, instance.ResolveVariables())
 	}
```

Comment:
Could it push the handling of implicit connectors into `connectorDef` also?

Reply:
Answered in the comment below